### ISSUE TITLE
使用Qwen1.5-SFT训练速度很慢

### ISSUE THREAD

**Issue by xschen-beb** [QwenLM/Qwen2#440]
使用以下的参数训练Qwen1.5-4b模型时训练速度很慢， log结果显示训练速度很慢而且初始的loss较大。然而使用qwen1.8b时初始的loss与训练时间为：
{'loss': 2.7494, 'learning_rate': 0.0, 'epoch': 0.01}

  0%|          | 1/340 [03:07<17:37:21, 187.14s/it]
  1%|          | 2/340 [06:09<17:19:45, 184.57s/it]

#### log信息：
[2024-05-27 15:57:18,800] [INFO] [real_accelerator.py:191:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-05-27 15:57:18,803] [INFO] [real_accelerator.py:191:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-05-27 15:57:18,807] [INFO] [real_accelerator.py:191:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-05-27 15:57:18,807] [INFO] [real_accelerator.py:191:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-05-27 15:57:32,558] [INFO] [comm.py:637:init_distributed] cdb=None
[2024-05-27 15:57:32,561] [INFO] [comm.py:637:init_distributed] cdb=None
[2024-05-27 15:57:32,564] [INFO] [comm.py:637:init_distributed] cdb=None
[2024-05-27 15:57:32,565] [INFO] [comm.py:637:init_distributed] cdb=None
[2024-05-27 15:57:32,565] [INFO] [comm.py:668:init_distributed] Initializing TorchBackend in DeepSpeed with backend nccl
[2024-05-27 15:57:37,127] [INFO] [partition_parameters.py:349:__exit__] finished initializing model - num_params = 483, num_elems = 3.95B
Loading data...
Formatting inputs...Skip in lazy mode
ninja: no work to do.
Time to load fused_adam op: 4.832568883895874 secondsTime to load fused_adam op: 4.831920146942139 secondsTime to load fused_adam op: 4.832811117172241 seconds

Time to load fused_adam op: 4.513275861740112 seconds
Parameter Offload: Total persistent parameters: 514560 in 201 params
{'loss': 17.2325, 'grad_norm': 266.94610078085555, 'learning_rate': 0.0, 'epoch': 0.0}
 0%|          | 1/940 [17:20<271:19:56, 1040.25s/it]

dependencies：
python==3.10
pytorch==2.2.0+cu118
transformers==4.41.0

parameters:
torchrun $DISTRIBUTED_ARGS finetune.py \
    --model_name_or_path $MODEL \
    --data_path $DATA \
    --bf16 True \
    --output_dir output_qwen/20240524/ \
    --num_train_epochs 4 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 8 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 100 \
    --save_total_limit 10 \
    --learning_rate 1e-5 \
    --weight_decay 0.1 \
    --adam_beta2 0.95 \
    --warmup_ratio 0.01 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --report_to "none" \
    --model_max_length 8192 \
    --gradient_checkpointing True \
    --lazy_preprocess True \
    --deepspeed finetune/ds_config_zero3.json

**Comment by jklj077** [comment]
Hi, 

It appears that you have conducted full-parameter finetuning on Qwen1.5-4B (1040s per batch) and Qwen1.5-1.8B (186s per batch). 

We don't currently benchmark the training speed as it may vary substantially with different options and devices. However, it could be possible that a 4B model requires more resources than a 1.8B model, especially if GPU memory (or bandwidth) is the bottleneck. 

Here are some options to try:
- check the memory usage and the gpu utilization rate to identify bottlenecks
- shorten the sequence length (--model_max_length)
- lower the batch size (--per_device_train_batch_size)
- enable flash attention v2 (via attn_implmenetation)
- disable gradient checkpointing if GPU memory is sufficient
- use other finetuning techniques, e.g., LoRA
